pkg/sources/circleci/circleci.go

```diff
@@ -229,34 +228,46 @@ func (s *Source) chunkAction(_ context.Context, proj project, bld build, act act
 		return err
 	}
 	defer res.Body.Close()
-	logOutput, err := io.ReadAll(res.Body)
-	if err != nil {
-		return err
-	}
 
 	linkURL := fmt.Sprintf("https://app.circleci.com/pipelines/%s/%s/%s/%d", proj.VCS, proj.Username, proj.RepoName, bld.BuildNum)
 
-	chunk := &sources.Chunk{
-		SourceType: s.Type(),
-		SourceName: s.name,
-		SourceID:   s.SourceID(),
-		Data:       removeCircleSha1Line(logOutput),
-		SourceMetadata: &source_metadatapb.MetaData{
-			Data: &source_metadatapb.MetaData_Circleci{
-				Circleci: &source_metadatapb.CircleCI{
-					VcsType:     proj.VCS,
-					Username:    proj.Username,
-					Repository:  proj.RepoName,
-					BuildNumber: int64(bld.BuildNum),
-					BuildStep:   stepName,
-					Link:        linkURL,
+	chunkReader := sources.NewChunkReader()
+	dataCh, errCh := chunkReader(ctx, res.Body)
+	for data := range dataCh {
+		chunk := &sources.Chunk{
+			SourceType: s.Type(),
+			SourceName: s.name,
+			SourceID:   s.SourceID(),
+			Data:       removeCircleSha1Line(data),
+			SourceMetadata: &source_metadatapb.MetaData{
+				Data: &source_metadatapb.MetaData_Circleci{
+					Circleci: &source_metadatapb.CircleCI{
+						VcsType:     proj.VCS,
+						Username:    proj.Username,
+						Repository:  proj.RepoName,
+						BuildNumber: int64(bld.BuildNum),
+						BuildStep:   stepName,
+						Link:        linkURL,
+					},
 				},
 			},
-		},
-		Verify: s.verify,
+			Verify: s.verify,
+		}
+		select {
+		case <-ctx.Done(): // priority to context cancellation
+			return ctx.Err()
+		default:
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case chunksChan <- chunk:
+			}
+		}
+	}
+	if err := <-errCh; err != nil {
+		ctx.Logger().Error(err, "Error reading chunk.")
+		return nil
 	}
```

Comment:
If there were no errors, we would read from a closed channel, which shouldn't block and would return the zero value. Therefore, this should be fine?

Reply:
I am considering whether it would be more manageable to return a unified type that includes both the []byte data and the error. For instance:
``` 
type ChunkResult struct {
  data []byte
  err error
}
```
The `ChunkResult` approach makes the code less complex and more intuitive to handle imo. By encapsulating both the []byte data and the error in a single object, it provides a clear and straightforward way to check for errors and access the data without having to manage multiple return values. This can lead to cleaner and easier to maintain code.

Thoughts?

Reply:
Yeah, I like the union approach. I can't imagine a large perf impact, though tests are better than imagination :)

My only recommendation is to use struct objects instead of pointers to avoid the indirection.